sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -1066,12 +1075,144 @@ async Task performProcessing()
+            // a fire-and-forget manner.  The processor does not assume responsibility for observing or surfacing exceptions
+            // that may occur in the handler, so the associated task is intentionally unobserved.
+
+            try
+            {
+                var connection = CreateConnection();
+                await using var connectionAwaiter = connection.ConfigureAwait(false);
+
+                var retryPolicy = Options.RetryOptions.ToRetryPolicy();
+                var cycleDuration = new Stopwatch();
+                var partitionIds = default(string[]);
+
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    cycleDuration.Restart();
+
+                    try
+                    {
+                        partitionIds = await connection.GetPartitionIdsAsync(retryPolicy, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception ex) when (ex.IsNotType<TaskCanceledException>())
+                    {
+                        // Logging for exceptions with the service operation are responsibility of the connection.
+
+                        _ = InvokeOnProcessingErrorAsync(ex, null, Resources.OperationGetPartitionIds, CancellationToken.None);
+                        partitionIds = default;
+                    }
+
+                    // Perform a load balancing cycle only in the case that partitions were successfully read.
+
+                    if (partitionIds != default)
+                    {
+                        var claimedOwnership = default(EventProcessorPartitionOwnership);
+
+                        try
+                        {
+                            claimedOwnership = await LoadBalancer.RunLoadBalancingAsync(partitionIds, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (EventHubsException ex)
+                            when (Resources.OperationClaimOwnership.Equals(ex.GetFailureOperation(), StringComparison.InvariantCultureIgnoreCase))
+                        {
+                            // If a specific partition was associated with the failure, it is carried by a well-known data member of the exception.
+
+                            var partitionId = ex.GetFailureData<string>();
+
+                            var partition = (partitionId ?? string.Empty) switch
+                            {
+                                string id when (id == string.Empty) => null,
+                                string _ when (ActivePartitionProcessors.TryGetValue(partitionId, out var partitionProcessor)) => partitionProcessor.Partition,
+                                _ => new TPartition { PartitionId = partitionId }
+                            };
+
+                            _ = InvokeOnProcessingErrorAsync(ex.InnerException ?? ex, partition, Resources.OperationClaimOwnership, CancellationToken.None);
```

Comment:
Chris moved a few things around so that we're reading from the `Data` property bag on the exception and not making assumptions about the message, which is intended as free-form text.   That's also the vehicle for sending the partition context along now.